src/cesium_ion/main.rb

```diff
@@ -0,0 +1,489 @@
+      'state' => state,
+      'response_type' => 'code',
+      'redirect_uri' => @@callbackUri
+    }
+
+    querystring = '';
+    query.each do |key, value|
+      querystring += (querystring.length == 0) ? '?' : amp
+      querystring += "#{key}=#{value}"
+    end
+
+    authorizeUrl = "#{@@baseServer}ion/oauth#{querystring}"
+    system("#{command} #{authorizeUrl}")
+
+    # Start a TCP server and block waiting for a single request
+    params = {}
+    begin
+      server = TCPServer.open(@@callbackHost, @@callbackPort)
+
+      socket = nil
+      count = 0
+      begin
+        sleep 1
+        socket = server.accept_nonblock
+      rescue IO::WaitReadable, Errno::EINTR
+        count += 1
+        if (count < 60) #Retry for a minute
+          retry
+        end
+      end
+
+      if socket.nil?
+        raise 'Connection Timeout'
+      end
+
+      line = socket.gets || ''
+      method, path = line.split
+      path ||= ''
+      path, query = path.split('?')
+      query ||= ''
+      while line = socket.gets
+        parts = line.split(' ', 2)
+        break if parts[0] == ""
+        if parts[0].downcase == 'host:'
+          host = parts[1].strip
+          break
+        end
+      end
+
+      query.split('&').each do |param|
+        key, value = param.split('=')
+        params[key] = value
+      end
+
+      if method != 'GET' || host.nil? || host != @@callbackDomain ||
+          path != @@callbackPath || params['state'] != state || params['code'].nil?
+          raise 'Invalid callback'
+      end
+    rescue => e
+      if @@debug
+        puts e
+      end
+    ensure
+      unless socket.nil?
+        socket.puts 'HTTP/1.1 200 OK'
+        socket.puts "Content-type: text/html"
+        socket.puts ''
+        socket.puts %{
+          <html>
+            <head>
+              <title>Authorization Complete</title>
+            </head>
+            <body>
+              Sketchup has been authorized to use your Cesium ion account.
+
+              You can close this window and go back to Sketchup.
+            </body>
+          </html>}
+
+        socket.close
+      end
+      server.close
+    end
+
+    # No code, so no point in moving forward
+    if params['code'].nil?
+      return
+    end
+
+    uri = URI.parse("#{@@apiServer}oauth/token")
+    http = get_http(uri)
+
+    request = Net::HTTP::Post.new(uri.request_uri)
+    tokenOptions = {
+      'client_id' => @@clientId,
+      'code' => params['code'],
+      'redirect_uri' => @@callbackUri,
+      'grant_type' => 'authorization_code'
+    }
+    request.body = tokenOptions.to_json
+    request['Content-Type'] = 'application/json'
+
+    response = http.request(request)
+
+    resultJson = {}
+    if response.code == '200' && !response.body.nil?
+      resultJson = JSON.parse(response.body)
+
+      File.write(tokenFile, resultJson['access_token'])
+    end
+
+    return resultJson['access_token']
+  end
+
+  ## END OAUTH2
+
+  ## EXPORT MODEL TO DISK
```

Comment:
Feature idea.  Since we're exporting to disk anyway, should we create a zipfile and upload that instead of each model individually?  That would save space in the user's account and also make the upload faster.

Reply:
If this isn't trivial, let me know and I'll write up an issue (but I bet it's trivial)